extensions/amp-ad/0.1/amp-ad.js

```diff
@@ -54,44 +55,48 @@ function copyAttributes(sourceElement, targetElement) {
 }
 
 export class AmpAd extends AMP.BaseElement {
-  constructor(element) {
-    super(element);
-  }
-
-  /** @override */
-  isLayoutSupported(layout) {
-    return isLayoutSizeDefined(layout);
-  }
 
   /** @override */
-  buildCallback() {
+  upgradeCallback() {
     const type = this.element.getAttribute('type');
     if (!type) {
       // Unspecified or empty type.  Nothing to do here except bail out.
-      return;
+      return null;
     }
-    let newChild;
     // TODO(tdrl): Check amp-ad registry to see if they have this already.
     if (!a4aRegistry[type] ||
         !a4aRegistry[type](this.getWin(), this.element)) {
       // Network either has not provided any A4A implementation or the
       // implementation exists, but has explicitly chosen not to handle this
       // tag as A4A.  Fall back to the 3p implementation.
-      newChild = this.element.ownerDocument.createElement(TAG_3P_IMPL);
-    } else {
-      // Note: The insertAmpExtensionScript method will pick the version number.
-      // If we ever reach a point at which there are different extensions with
-      // different version numbers at play simultaneously, we'll have to make sure
-      // that the loader can handle the case.
-      const extensionTag = networkImplementationTag(type);
-      newChild = this.element.ownerDocument.createElement(extensionTag);
-      /*OK*/insertAmpExtensionScript(this.getWin(), extensionTag, true);
+      return new AmpAd3PImpl(this.element);
```

Comment:
@tdrl 

> I take it that we'll be able to upgrade to arbitrary ad network-provided implementations in the same way?

Yes. I'm not 100% yet sure how this will be accomplished, but this is definitely the way to go. The callback returns promise, so any async way will do.

> I've had trouble in the past with things getting called before attributes are set and failing because they couldn't see type.

Yes. This was the secondary reason. This is why I deferred this call to first attachment in DOM, which is most of time will be shortly before `buildCallback` would have been called.

> I assume this will handle replaying any events that arrived at the <amp-ad> element before it is upgraded to the post-upgrade element?

Yes. It basically takes advantage of the existing "upgrade" system which already works as expected. So this should work exactly the same.